File: src/main/java/com/tmvkrpxl0/tcombat/common/container/MobileDispenserGui.java
package com.tmvkrpxl0.tcombat.common.container;

import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.entity.player.PlayerInventory;
import net.minecraft.inventory.container.Container;
import net.minecraft.inventory.container.DispenserContainer;
import net.minecraft.inventory.container.INamedContainerProvider;
import net.minecraft.inventory.container.Slot;
import net.minecraft.item.ItemStack;
import net.minecraft.util.text.ITextComponent;
import net.minecraftforge.items.CapabilityItemHandler;
import net.minecraftforge.items.IItemHandler;
import net.minecraftforge.items.ItemStackHandler;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.util.Optional;

public class MobileDispenserGui implements INamedContainerProvider {
    private final ItemStack mobileDispenser;

    public MobileDispenserGui(ItemStack mobileDispenser) {
        this.mobileDispenser = mobileDispenser;
    }

    @Nonnull
    @Override
    public ITextComponent getDisplayName() {
        return mobileDispenser.getHoverName();
    }

    @Nullable
    @Override
    public Container createMenu(int windowId, @Nonnull PlayerInventory inventory, @Nonnull PlayerEntity player) {
        Optional<IItemHandler> optional = mobileDispenser.getCapability(CapabilityItemHandler.ITEM_HANDLER_CAPABILITY).resolve();
        DispenserContainer container = new DispenserContainer(windowId, inventory){
            @Override
            public void removed(@Nonnull PlayerEntity playerIn) {
                if(optional.isPresent() && player.isEffectiveAi()){
                    ItemStackHandler handler = (ItemStackHandler) optional.get();
                    for(int i = 0;i<9;i++){
                        Slot slot = this.getSlot(i);
                        ItemStack stack = slot.getItem();
                        handler.setStackInSlot(i, stack);
                    }
                }
                super.removed(playerIn);
            }
        };
        if(optional.isPresent() && player.isEffectiveAi()){
            IItemHandler handler = optional.get();
            for(int i = 0;i<9;i++){
                container.setItem(i, handler.getStackInSlot(i));
            }
        }
        return container;
    }

}

File: src/main/kotlin/com/tmvkrpxl0/tcombat/common/container/MobileDispenserGui.kt
package com.tmvkrpxl0.tcombat.common.container

import net.minecraft.entity.player.PlayerEntity
import net.minecraft.entity.player.PlayerInventory
import net.minecraft.inventory.container.Container
import net.minecraft.inventory.container.DispenserContainer
import net.minecraft.inventory.container.INamedContainerProvider
import net.minecraft.item.ItemStack
import net.minecraft.util.text.ITextComponent
import net.minecraftforge.items.CapabilityItemHandler
import net.minecraftforge.items.ItemStackHandler
import javax.annotation.Nonnull

class MobileDispenserGui(private val mobileDispenser: ItemStack) : INamedContainerProvider {
    @Nonnull
    override fun getDisplayName(): ITextComponent {
        return mobileDispenser.displayName
    }

    override fun createMenu(windowId: Int, @Nonnull inventory: PlayerInventory, @Nonnull player: PlayerEntity): Container {
        val optional = mobileDispenser.getCapability(CapabilityItemHandler.ITEM_HANDLER_CAPABILITY).resolve()
        val container: DispenserContainer = object : DispenserContainer(windowId, inventory) {
            override fun removed(@Nonnull playerIn: PlayerEntity) {
                if (optional.isPresent && !player.level.isClientSide) {
                    val handler = optional.get() as ItemStackHandler
                    for (i in 0..8) {
                        val slot = getSlot(i)
                        val stack = slot.item
                        handler.setStackInSlot(i, stack)
                    }
                }
                super.removed(playerIn)
            }
        }
        if (optional.isPresent && !player.level.isClientSide) {
            val handler = optional.get()
            for (i in 0..8) {
                container.setItem(i, handler.getStackInSlot(i))
            }
        }
        return container
    }
}